### PR TITLE
[MBL-15819][Teacher] Comment Library

### PR DESCRIPTION
**Pull request by tamaskozmer**
refs: MBL-15819
affects: Teacher
release note: Added comment library to Speed Grader.

Test plan:
- Most of the use cases are covered in the specification linked in the ticket.
- Also verify that sending a comment or adding an attachment when the comment library is open closes the comment library.

### CODE REVIEW

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by kdeakinstructure** [comment]
QA 👍

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

File: apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/CommentLibraryViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                    val commentLibraryEnabled = userSettings.commentLibrarySuggestions
+                    teacherPrefs.commentLibraryEnabled = commentLibraryEnabled
+                    commentLibraryEnabled
+                }
+
+                if (commentLibraryEnabled) {
+                    loadCommentLibraryContent()
+                }
+            } catch (e: Exception) {
+                // No-op Silently fail if we don't have info about the comment library, and just don't show anything.
```
Comment (hermannakos) [inline apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/CommentLibraryViewModel.kt:80]: Just for debugging purposes we may want to log the exception

File: apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/Highlighter.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import android.graphics.Typeface
+import android.text.Spannable
+import android.text.SpannableString
+import android.text.style.StyleSpan
+
+object Highlighter {
+
+    @JvmStatic
+    fun createHighlightedText(data: HighlightedSuggestionViewData): Spannable {
```
Comment (hermannakos) [inline apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/Highlighter.kt:11]: We could use this highlighter somewhere else as well. Could we just pass the required params to it so it's not restricted to the `HighlightedSuggestionViewData`

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

File: apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/CommentLibraryViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    private fun updateFilteredSuggestions(currentSubmissionId: Long) {
+        val commentLiveData = commentsBySubmission.computeIfAbsent(currentSubmissionId) { MutableLiveData() }
+        val query = commentLiveData.value?.comment ?: ""
+        filterSuggestions(query)
+    }
+
+    private fun filterSuggestions(query: String) {
+        val filteredSuggestions = allSuggestions
+            .filter { it.contains(query, ignoreCase = true) }
```
Comment (hermannakos) [inline apps/teacher/src/main/java/com/instructure/teacher/features/speedgrader/commentlibrary/CommentLibraryViewModel.kt:130]: We should also ignore accented characters as well.

**Review by hermannakos — CHANGES_REQUESTED** [review]
LGTM! My only real RC is about the filtering.

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by hermannakos — APPROVED** [review]

**Pull request merged**